### PR TITLE
cilium-cli: Show config.cilium.io annotations on configmap

### PR DESCRIPTION
**Pull request by joamaki**
Add support for showing config.cilium.io annotations in configmaps/cilium-config in "cilium status". This allows informing the user of a non-fatal misconfigured or unsupported configuration.

Example output with annotation {"config.cilium.io/example": "This is an example warning"}:
```
  $ cilium status
  ...
  Configuration:                This is an example warning
```

### CODE REVIEW

File: cilium-cli/status/k8s.go [hunk 1]
```diff
@@ -308,6 +309,23 @@ func (k *K8sStatusCollector) podStatus(ctx context.Context, status *Status, name
 	return nil
 }
 
+func (k *K8sStatusCollector) ciliumNodeAnnotations(ctx context.Context, status *Status) error {
+	nodes, err := k.client.ListCiliumNodes(ctx)
```
Comment (joamaki) [inline cilium-cli/status/k8s.go]: Instead of pulling the whole CiliumNode we could pull only the `PartialObjectMetadata` if we set the `Accept` header accordingly (like in `pkg/k8s/synced/crd.go`). It would make this significantly more complicated, but would save a bit of time and transmitted bytes. Not sure if it's worth it.

Comment (joamaki) [reply cilium-cli/status/k8s.go]: Was curious, so implemented it:
```

func (c *Client) ListCiliumNodeAnnotations(ctx context.Context) (map[string]map[string]string, error) {
	const pomListHeader = "application/json;as=PartialObjectMetadataList;v=v1;g=meta.k8s.io,application/json;as=PartialObjectMetadataList;v=v1beta1;g=meta.k8s.io,application/json"
	rest := c.CiliumClientset.CiliumV2().RESTClient()
	opts := metav1.ListOptions{}
	list := metav1.PartialObjectMetadataList{}
	err := rest.Get().
		Name("ciliumnodes").
		SetHeader("Accept", pomListHeader).
		VersionedParams(&opts, metav1.ParameterCodec).
		Do(context.TODO()).
		Into(&list)
	if err != nil {
		return nil, err
	}
	annotations := map[string]map[string]string{}
	for _, meta := range list.Items {
		annotations[meta.Name] = meta.Annotations
	}
	return annotations, nil
}
```

yay or nay?

File: cilium-cli/status/status.go [hunk 2]
```diff
@@ -433,11 +441,44 @@ func (s *Status) Format() string {
...
 	}
 
+	header = "Taints:"
+	for taint, nodes := range getCiliumNodeTaints(s.NodeAnnotations) {
+		fmt.Fprintf(w, "%s\t%s\t%s\n", header, defaults.AgentDaemonSetName, trimNodeNames(nodes))
+		for _, line := range strings.Split(taint, "\n") {
+			fmt.Fprintf(w, "\t\t%s\n", line)
+		}
+		header = ""
+	}
```
Comment (tklauser) [inline cilium-cli/status/status.go]: Would be nice to display these sorted alphabetically, like e.g. errors and warnings above. Otherwise these lines will have a different sorting order each time `cilium status` is invoked or during `cilium status --wait` which makes it a bit hard to follow what's going on. Also see https://github.com/cilium/cilium/pull/34927 for context.

Comment (joamaki) [reply cilium-cli/status/status.go]: Ah good catch, thanks!

**Review by tklauser — APPROVED** [review]
Neat, thanks!

File: pkg/annotation/k8s.go [hunk 3]
```diff
@@ -37,6 +37,9 @@ const (
 	// CNIPrefix is the common prefix for CNI related annotations.
 	CNIPrefix = "cni.cilium.io"
 
+	// TaintPrefix is the common prefix for taint annotations.
+	TaintPrefix = "taint.cilium.io"
```
Comment (joestringer) [inline pkg/annotation/k8s.go]: This is a neat facility, though I am concerned the use of the name "taint", given that we commonly operate in k8s environments and kubernetes already has a taints concept that impacts scheduling: https://kubernetes.io/docs/concepts/scheduling-eviction/taint-and-toleration/ . Perhaps using something like `status.cilium.io` or reusing `config.cilium.io` could be viable alternatives?

Comment (joamaki) [reply pkg/annotation/k8s.go]: Yep makes sense. I'll use a different prefix and different header in cilium status to avoid the confusion.

File: install/kubernetes/cilium/templates/cilium-operator/clusterrole.yaml [hunk 4]
```diff
@@ -18,6 +18,13 @@ rules:
   - get
   - list
   - watch
+- apiGroups:
+  - ""
+  resources:
+  - configmaps
```
Comment (gandro) [inline install/kubernetes/cilium/templates/cilium-operator/clusterrole.yaml:24]: Can we add `resourceNames` here such that only access to the `cilium-config` configmap is provided?

Comment (joamaki) [reply install/kubernetes/cilium/templates/cilium-operator/clusterrole.yaml:24]: thanks, added!

**Review by gandro — APPROVED** [review]

File: install/kubernetes/cilium/templates/cilium-operator/clusterrole.yaml [hunk 5]
```diff
@@ -18,6 +18,15 @@ rules:
...
+  resourceNames:
+  - cilium-config
+  verbs:
+   # allow patching of the configmap to set annotations
```
Comment (asauber) [inline install/kubernetes/cilium/templates/cilium-operator/clusterrole.yaml:28]: very nit: there is an extra leading space here

**Review by asauber — APPROVED** [review]
